### PR TITLE
Change graph generation of sender and recipient graphs to Chart.js

### LINKED ISSUE #283: Bug in rep_top_viruses.php
Legends not well printed.

pict3
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/37354821-bug-in-rep_top_viruses-php?utm_campaign=plugin&utm_content=tracker%2F585686&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F585686&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Skywalker-11**
This will change the generation of the graphes to [Chart.js](http://www.chartjs.org/docs/#getting-started-download-chart-js) (MIT License)

This uses HTML5 canvas objects to generate the graphes on the client side and so doesn't require the php library JPGraph, php-gd and no writable cache directory on the server. It should be supported by all current browsers.

Includes PR #474
Fixes #283 and #476

### CODE REVIEW

**Comment by stefaweb** [comment]
Great!

And #323 ?

'errorcachedirnotwritable03' => 'Erreur : impossible d'écrire dans le répertoire de cache %s',

**Comment by Skywalker-11** [comment]
#323 is another issue which is related to how we identify the used virus scanner. 

Also I just adjusted the rep_top_recipient* and rep_top_sender* files which are using the refactored graph generation. If this is acceptable to @endelwar I will do it for the other reps

**Comment by stefaweb** [comment]
ok

**Comment by Skywalker-11** [comment]
the last commits gives the ability to use the function for graph generation for other report layouts

**Comment by stefaweb** [comment]
A screen shot of the new graph?

**Comment by Skywalker-11** [comment]
image

**Comment by stefaweb** [comment]
If I can afford advice.

Try to reuse the colors defined in the jpgraph version. I spent time to select colors for best view.

For sizes, forms and legends placement, try to mimic the jpgraph version. And in a second time, solve all problem of covering.

**Comment by endelwar** [comment]
Need to test this, but looks promising, great job

**Comment by endelwar** [comment]
I'm going to merge this after 1.2.0 release (I think already in 1.2.1)

**Comment by stefaweb** [comment]
@Skywalker-11 

When you have time (not urgent) can you sync this PR. I'll to try it.

**Comment by Skywalker-11** [comment]
To rebase with develop I took develop branch and applied the changes of this pr manually so we get a clean commit history

**Comment by endelwar** [comment]
@Skywalker-11 can you rebase this to 1.2.0 so that I can include it in 1.2.1?
also please move the js library in lib directory, and put an empty index.html file in Chart.js/ dir

**Comment by Skywalker-11** [comment]
@endelwar is done.

**Comment by stefaweb** [comment]
I prefer the old graph.
Can you wait? 
I would like to work on the graph design a little before.

File: mailscanner/functions.php [hunk 1]
```diff
@@ -3895,73 +3916,93 @@ function ($val) {
...
-        }
-    } else {
-        echo '<tr><td class="center">' . sprintf(__('errorcachedirnotwritable03'), CACHE_DIR) . '</td></tr>';
-    }
-
-    echo '<tr>';
+    //create canvas graph
+    $bgcolors = getHexColors(count($data[$graphColumn['dataColumn']]));
+    echo '<canvas id="reportChart" class="reportGraph"></canvas>
+  <script src="js/Chart.js/Chart.min.js"></script>
```
Comment (endelwar) [inline mailscanner/functions.php]: no more in js dir

File: mailscanner/functions.php [hunk 2]
```diff
@@ -3895,73 +3916,93 @@ function ($val) {
...
     for ($i = 0; $i < $numResult; $i++) {
         echo '    <tr style="background-color: #EBEBEB">' . "\n";
-        foreach ($sqlColumns as $sqlColumn) {
-            if (isset($valueConversions[$sqlColumn])) {
-                echo '     <td>' . $data[$sqlColumn . 'conv'][$i] . '</td>' . "\n";
-            } else {
-                echo '     <td>' . $data[$sqlColumn][$i] . '</td>' . "\n";
-            }
+        foreach ($columns as $columnName => $columnTitle) {
+            echo '     <td>' . $data[$columnName][$i] . '</td>' . "\n";
```
Comment (endelwar) [inline mailscanner/functions.php]: here is missing formatting of numbers and conversion of size from bytes to human readable units

Comment (Skywalker-11) [inline mailscanner/functions.php]: When rebasing I forgot to change adjust the column names. Is fix now

**Comment by stefaweb** [comment]
Can we do this with new graph code?

top_senders_by_quantity png1490171794

And, the title font size  in new graph is too small.

**Comment by Skywalker-11** [comment]
Not sure about that but you can basically change the library to everything that accepts json. So if you find an other library that can do it tell me.

But I will look at Graph.js again

**Comment by stefaweb** [comment]
I see theses 'problems'.

- title font size in new graph is too small. 
- graph too large/big on the page
- when you have many legends on top, its not easy to read
- we do not see % when we print the graph page. We only see them when we move the mouse on.

**Comment by Skywalker-11** [comment]
font size of title can be changed by using 
```
options: {
  title: {
    display: true,
    text: "' . $reportTitle .'",
    fontSize: 18 //or whatever size it should be
  }
}
```

graph size can be changed by adjusting min-height, min-width, max-height, max-width, height, width for   `.reportGraph` in style.css

**Comment by stefaweb** [comment]
@Skywalker-11 

Can you  rebase with last develop version by the end of the day? 
Will try to find some spare time this week-end to look a the last version of graph.

**Comment by Skywalker-11** [comment]
Current develop is merged. I will do a proper rebase on the weekend

**Comment by stefaweb** [comment]
I tried the last version with last develop and go this:

capture d ecran 2017-03-26 a 11 05 58

**Comment by stefaweb** [comment]
I've tried this but do not understand how align legend and pie.

Seems 3D pie is not available. Not found how to print % around the pie.

```
      options: {
        title: {
          display: true,
          text: "' . $reportTitle . '",
          fontSize: 18
        },
        legend: {
          display: true,
          position: "left",
          fullWidth: false
        },
        responsive: false
      }

```

capture d ecran 2017-03-26 a 11 37 42

**Comment by stefaweb** [comment]
For reference.

https://d3js.org

50 best JavaScript charting libraries
http://www.cssauthor.com/javascript-charting-libraries/

**Comment by Skywalker-11** [comment]
@stefaweb are you sure you had the latest version? There should already be percentages in the pieces. I don't see that Notices either. But I just noticed that the position of the graph seems to be buggy so that when you have large legends the graph will not be visible

I rebased it again with current develop

**Comment by stefaweb** [comment]
Found. My git directory was not uptodate.

Added in functions.php -> 
title -> fontSize: 18
legend -> position: "left"
.reportGraph -> height: 750px; and width: 750px;  (seems 850 x 350 doesn't work)

But I still have the problem to align legend and pie.

And the % overflows pie border.

capture-d ecran-2017-03-26-a-13 38 37

**Comment by Skywalker-11** [comment]
Here's another try with 3dpie.js based on d3js [https://github.com/Skywalker-11/1.2.0/tree/feature/3djs](https://github.com/Skywalker-11/1.2.0/tree/feature/3djs)

**Comment by stefaweb** [comment]
Have a look on Highcharts. Seems very nice. Work with JSON and seems have the features we need (same and more as in JPGraph).

http://api.highcharts.com/highcharts

**Comment by Skywalker-11** [comment]
placing the percentage always inside is a bit tricky:
the `-10` at var radius has to be reduced and then the offset (-8 or -5) in ctx.fillText has to be adjusted 
```
  var radius = model.outerRadius-10; //where to place the text around the center
            var x = Math.sin((curr+part/2)*2*Math.PI)*radius;
            var y = Math.cos((curr+part/2)*2*Math.PI)*radius;
            ctx.fillText((part*100).toFixed(0)+"%",model.x + x - 8 , model.y - y - 5 );
```

I don't know how I could reduce that space.

**Comment by stefaweb** [comment]
> I don't know how I could reduce that space.

I've read most of the Chart.js API, it do not have the needed feature. DOT. We need to try another one (d3js or Highcharts).

**Comment by Skywalker-11** [comment]
Heres a highchart version [https://github.com/Skywalker-11/1.2.0/blob/feature/highchart/mailscanner/functions.php](https://github.com/Skywalker-11/1.2.0/blob/feature/highchart/mailscanner/functions.php).

I prefer this the most at the moment because it properly can add line breaks to the labels. Only problem is if we have many entries with long labels then they overlap eg. [http://jsfiddle.net/hmn6z9bf/2/](http://jsfiddle.net/hmn6z9bf/2/)

**Comment by stefaweb** [comment]
I like too. 

highchart have the function I was looking for. A line to estrange the pie from the legend or from the %.

I'm sure we can adjust all these stuff.

**Comment by stefaweb** [comment]
Test.

capture d ecran 2017-03-26 a 15 39 29

Can you do a test with rep_total_mail_by_date.php ?

**Comment by stefaweb** [comment]
If possible.

We can have legend as in the graph test for:

Top Sender Domains by Quantity
Top Sender Domains by Volume
Top Recipient Domains by Quantity
Top Recipient Domains by Volume

% inside the pie or at the start of legend.

And legends below the pie stacked in 2 columns for:

Top Senders by Quantity
Top Senders by Volume
Top Recipients by Quantity
Top Recipients by Volume

With this solution no more problem wth very very long email.

And add % at the end of the line going to the pie.

**Comment by stefaweb** [comment]
Shit.

at the right button.

capture d ecran 2017-03-26 a 16 10 10

@endelwar will not agree with that?

**Comment by Skywalker-11** [comment]
rep_total_mail_by_date is not using a pie graph but a bar graph. I haven't looked at that type and haven't refactored it till yet either.

This adds a separate legend but consumes the place for graph... [https://github.com/Skywalker-11/1.2.0/commit/2cbc61d9075d6c9772cba46b46e202315266966f](https://github.com/Skywalker-11/1.2.0/commit/2cbc61d9075d6c9772cba46b46e202315266966f)

**Comment by stefaweb** [comment]
We can use 780 pixel for the length of the graph size. Height can be around 350 pixels.

Problem with printing (stacked to the left, small size).

capture d ecran 2017-03-26 a 16 25 26

The print is good quality.

**Comment by shawniverson** [comment]
Is the highcharts licensing compatible with GPL?

It looks like it is not free.

**Comment by stefaweb** [comment]
Greuuu!

We also have highcharts.com printed on the "exports" (pdf, png,...)

Not GPL. Free of use if not commercial only.

**Comment by stefaweb** [comment]
More details.

```
Do you want to use Highcharts for a personal website, a school site or a non-profit organisation? Then you don't need the author's permission, just go on and use Highcharts.

One of the key features of Highcharts is that under any of the licenses, free or not, you are allowed to download the source code and make your own edits. This allows for personal modifications and a great flexibility.

```

We can edit the code, but remove @ from screen en prints? Not sure at all.

**Comment by Skywalker-11** [comment]
Problem is if someone want to sell mailwatch as a service to a customer. Then they can't use highcharts

**Comment by stefaweb** [comment]
@Skywalker-11 : do you have a screen shot of the d3js test?

**Comment by Skywalker-11** [comment]
image

[http://jsfiddle.net/9xcqp01n/](http://jsfiddle.net/9xcqp01n/)

**Comment by Skywalker-11** [comment]
The labels that are too long will be shorten by `...`. The api 3dpie that I'm using does not support seperate labels yet but it's a planned feature

**Comment by stefaweb** [comment]
Another Open Source candidate.

Built with D3.js and Ember.js. On GitHub and living project.

http://opensource.addepar.com/ember-charts/#/pie
https://github.com/Addepar/ember-charts

**Comment by stefaweb** [comment]
@Skywalker-11 

Look a rgraph. 

Seems very powerful. We can have 3D for pie and others graphs too. JSON and AJAX compatible.

https://www.rgraph.net/demos/svg-pie-json.html
https://www.rgraph.net/demos/pie-3d.html
https://www.rgraph.net/demos/svg-bar-3d.html
https://www.rgraph.net/demos/bar-3d-effect-multi-row.html

**Comment by endelwar** [comment]
highcharts is a no go because its licence

> May I use your Software under the Non-Commercial License for Open Source Projects?
> 
> Although Highsoft´s Software have open source codes, our software is not licensed as an open source software and are unfortunately not compatible with any open source software license like Apache 2 or any GPL. See also Non-Commercial Redistribution

**Comment by stefaweb** [comment]
>  highcharts is a no go because its licence

I agree.

What do think of rgraph?

**Comment by Skywalker-11** [comment]
The default pie of rgraph is not good: 
image

Heres the branch with the corresponding code: [https://github.com/Skywalker-11/1.2.0/tree/feature/rgraph](https://github.com/Skywalker-11/1.2.0/tree/feature/rgraph)

**Comment by stefaweb** [comment]
No way with lines. We need to stay as in jpgraph version...

rgraph has the advantage of being able to create 3D graphics, which is not the case with other libraries we tested (except the non free highcharts).

3f66423c-0ee4-11e7-9135-2048b0c3d300

**Comment by Skywalker-11** [comment]
But do we realy need 3D?

**Comment by shawniverson** [comment]
3D not necessary, but makes me look cool 😎

**Comment by stefaweb** [comment]
Not having 3D graphics is a "regression" compared to jpgraph. It is also prettier and easier to read.

**Comment by Skywalker-11** [comment]
An other try but I currently haven't found a way on how I prevent overlapping of the values and not having that sticks
image

**Comment by endelwar** [comment]
In a project that I'm developing I'm testing this chart.js plugin to add percentage inside pie charts: https://github.com/emn178/Chart.PieceLabel.js
It works fine if the pie slice can contains the numbers

**Comment by Skywalker-11** [comment]
How shall we precede with this?

**Comment by endelwar** [comment]
I think that there isn't any ready library that can fit all graph cases, especially the ones that have very small slices (2% or less).
Let's merge this PR as is (with a `git rebase develop` before merging) and release the feature, and then work on it for optimization

**Pull request merged**